src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java

```diff
@@ -290,7 +291,14 @@ public ResponseEntity<byte[]> getFileUploadSubmission(@PathVariable Long exercis
             throw new AccessForbiddenException();
         }
 
-        return buildFileResponse(filePathService.actualPathForPublicPath(URI.create(submission.getFilePath())), false);
+        String filePath = submission.getFilePaths().stream().filter(path -> path.endsWith(fileName)).findFirst().orElse(null);
```

Comment:
Have you tested what would happen if a student uploads two different files with the same name? Are they correctly saved and retrieved?

Reply:
I suppose it’s quite the edge-case. If it is hard to implement, I guess forbidding multiple files having identical names would be okay, too.

Reply:
The most obvious solution which I've already implemented is to compute a hash of the file contents when they are staged in the stage component.
To continue displaying the correct filename without the hash prefix, you just modify the file details function to cut off the hash before displaying it.
That would not require any more testing on the server and client side since the change is isolated to two files.

Only problem is the database migration can't compute the hash.

Even more promising would be to prefix the file with its index in the list instead of a hash but that would require a small change to the server.
I'm looking into that now especially because you might be able to notice the delay caused by the hash function on some slow clients.

Long term you'd probably want to save the name in a seperate column and just refer to files by /api/path/{hash}.{extension} instead of /api/path/{name}.{extension} because that would leave the spaces, special chars etc in the names intact. But that would require much more significant change and should probably be a seperate pr at some point.

Reply:
Nevermind the index version should work. I got rid of the hash